apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/TimedBufferingConsumer.cs

```diff
@@ -21,56 +21,50 @@
 namespace Google.Cloud.Diagnostics.Common
 {
     /// <summary>
-    /// A <see cref="IFlushableConsumer{T}"/> that will flush the buffer during a receive
-    /// call if the given amount of time has passed..
+    /// A <see cref="IFlushableConsumer{T}"/> that will automatically flush the buffer after a 
+    /// given amount of time.
     /// </summary>
     internal class TimedBufferingConsumer<T> : FlushableConsumerBase<T>
     {
         /// <summary>The consumer to flush to.</summary>
         private readonly IConsumer<T> _consumer;
 
-        /// <summary>The minimum amount of time to wait between automatically flushing the buffer.</summary>
-        private readonly TimeSpan _waitTime;
-
-        /// <summary>A clock for getting the current timestamp.</summary>
-        private readonly IClock _clock;
-
         /// <summary>
         /// The buffered items. This is not readonly as it is replaced when the buffer is flushed.
         /// </summary>
         private List<T> _items = new List<T>();
 
-        /// <summary>The earliest time of the next automatica flush.</summary>
-        private DateTime _nextFlush;
+        /// <summary>The timer to automatically flush the buffer.</summary>
+        private IThreadingTimer _timer;
 
-        private TimedBufferingConsumer(IConsumer<T> consumer, TimeSpan waitTime, IClock clock)
+        internal TimedBufferingConsumer(IConsumer<T> consumer, TimeSpan waitTime, IThreadingTimer timer)
         {
             _consumer = GaxPreconditions.CheckNotNull(consumer, nameof(consumer));
-            _waitTime = waitTime;
-            _clock = GaxPreconditions.CheckNotNull(clock, nameof(clock));
-            _nextFlush = _clock.GetCurrentDateTimeUtc().Add(_waitTime);
+            _timer = GaxPreconditions.CheckNotNull(timer, nameof(timer));
+            // Initialize the timer to flush ever wait time interval.
+            _timer.Initialize((e) => { Flush(); }, waitTime);
+        }
+
+        ~TimedBufferingConsumer()
+        {
+            Flush();
+            _timer.Dispose();
```

Comment:
[This section](https://msdn.microsoft.com/en-us/library/b1yfkh5e(v=vs.110).aspx#basic_pattern) of the link above shows how to properly dispose of things. Basically, if someone calls Dispose directly, you want to dispose everything you own that's disposable or needs cleanup. You also want to prevent the finalizer from running. If instead the finalizer is run, you only want to cleanup unmanaged resources. Other managed, disposable objects will have their finalizers run as well and that might happen before the current object is disposed.